base/serialize.jl

```diff
@@ -770,11 +835,48 @@ function deserialize(s::AbstractSerializer, ::Type{GlobalRef})
     end
 end
 
+function deserialize_wrapper_datatype(s::AbstractSerializer)
+    number = read(s.io, UInt64)
+    if haskey(known_object_data, number)
+        ndt = known_object_data[number]::DataType
+        makenew = false
+    else
+        ndt = ccall(:jl_new_uninitialized_datatype, Any, ())::DataType
+        makenew = true
+        ndt.super = Any
+    end
+    resolve_ref_immediately(s, ndt)
+
+    # deserialize_typename can assume unwrap_unionall(wrapper).parameters is set
```

Comment:
The code as written doesn't directly assume that --- i.e. `deserialize_typename` itself doesn't look at that info. We just need to make sure the type isn't used for anything serious until it's initialized. Hopefully null pointer accesses would indicate if we violate that. The boolean flags could be initialized earlier though, at least.